Switch to SCons

micahswitzer opened the pull request:
It's time.

Closes #324

micahswitzer committed:
build: Add a basic SConstruct that can build kernel cpp files

Kfeavel commented on SConstruct:
```diff
@@ -0,0 +1,55 @@
+        '-fpermissive', '-fno-rtti', '-fno-exceptions', '-fno-use-cxa-atexit',
+    ],
+    CPPDEFINES={
+        'REPO_URL': '\\"$REPO_URL\\"',
```
We might want to use `shlex.quote()` for these. Just a suggestion, take it or leave it.

micahswitzer replied:
Ooh yeah, good idea. I already had some interesting ideas for passing through these build variables as defines.

Kfeavel replied:
I tried using `shlex` and wasn't able to get it to work. I think it might have something to do with the fact that we're trying to access variables that aren't yet "commited" to the environment.

micahswitzer replied:
Hmm yeah, I guess I'll have to play with it a bit. I think it might be possible with a little bit of creativity. Thinking about it a little more I don't think it's really necessary since there shouldn't be any adversarial input to the build system, but I'm not opposed to using it if we can get it to work.

micahswitzer committed:
build: Split SConscript into SConstruct and build assembly files

micahswitzer committed:
build: Link kernel with SCons

Kfeavel commented:
When you add the `run` task back in like we have in the `Makefile`, can you add the following to the Qemu options?

`-monitor telnet:127.0.0.1:1234,server,nowait`

micahswitzer committed:
build: Add basic echFS SCons builder

Kfeavel committed:
Specify `ranlib` binary in SConscript
Fixes issues when compiling without `xyris-build`

Kfeavel committed:
Code style update and add color to output

Kfeavel committed:
Add warnings to SConstruct

Kfeavel committed:
Replace `site_scons` with `scones` directory
It's Thanksgiving and I thought calling the scons tools scones
was a funny idea. Personally I don't like the `site_scons` directory
naming anways.

Kfeavel committed:
Install kernel to `dist`

Kfeavel committed:
Add missing `-lgcc` and `-mno-avx` flags

Kfeavel committed:
Rename `x86` to `i686` in SConstruct

Kfeavel committed:
Fix echfs scons tool issue with macOS

Kfeavel commented:
Remaining tasks:

- [ ] Install Limine to echFS image (create Limine SCons tool)
- [ ] Create Limine ISO
- [ ] Build & run unit tests
- [ ] Get git commit short hash

Kfeavel commented:
Is there a way to make targets that the user can invoke like in the Makefile? We need a way to run the kernel in Qemu, so if we can't find a way to do that with SCons, we'll likely need to keep the Makefile.

Kfeavel committed:
Update capitalization to match `dev` branch

micahswitzer commented:
> Is there a way to make targets that the user can invoke like in the Makefile? We need a way to run the kernel in Qemu, so if we can't find a way to do that with SCons, we'll likely need to keep the Makefile.

Yeah that should be possible with a `Command`

micahswitzer committed:
build: Update to latest dev

micahswitzer committed:
test: Allow Catch2 download errors to be bubbled up

micahswitzer committed:
build: WIP SCons build full image

Kfeavel committed:
Fix catch2 header download issues

Kfeavel committed:
WIP: Generate bootable ext2 image with SCons

Kfeavel committed:
Update SConstruct with variables from Makefile

Kfeavel committed:
Build bootable ext2 image with SCons

Kfeavel committed:
Get current git commit for CPP `COMMIT` define

Kfeavel committed:
fixup

Kfeavel committed:
WIP: Compile unit tests with SCons

Kfeavel committed:
Compile unit tests using SCons

Kfeavel committed:
Split arch src and common src in SConscript

Kfeavel committed:
Use glob function in unit test source

Kfeavel committed:
Add `i686_elf.py` SCons tool

Kfeavel committed:
Add colors SCons tool

Kfeavel committed:
Set kernel targets as default SCons targets
Perform cleanup on SConstruct

Kfeavel committed:
Use `.append()` instead of `+=`

Kfeavel committed:
Update Scuba aliases to use SCons

Kfeavel committed:
Create mode specific build target lists

Kfeavel committed:
Add phony `qemu` run targets to SConstruct

Kfeavel committed:
Add `doxygen` SCons tool

Kfeavel committed:
Remove all makefiles

Kfeavel committed:
Fix missing unit test install step

Kfeavel committed:
Update CI artifact names to reflect SCons output

Kfeavel committed:
Update CI to parallelize tests and kernel builds

Kfeavel commented:
Before merging, the `liballoc` repo should have the `scons` branch merged in and this branch update the submodule reference to `xyris` (with the newly merged `SConscript`)

Kfeavel committed:
Unite `glob` functions in `glob.py` SCons tool

Kfeavel committed:
Update liballoc to latest `xyris` branch commit
Adds `SConscript` to `xyris` branch

Kfeavel committed:
Begin refactoring `ext2.py` Scons tool
Builders and construction vars should be used instead of formatted strs

micahswitzer commented on Scones/glob.py:
```diff
@@ -0,0 +1,34 @@
+    sources = []
+    for (dirpath, dirs, files) in os.walk(root):
+        for extension in extensions:
+            sources.append(env.Glob(dirpath + '/*' + extension))
```
This should have slightly better performance
```suggestion
            sources.append(env.Glob(f'{dirpath}/*{extension}'))
```

Kfeavel replied:
Addressed in a9bfd261b36f4f272ff1412e79cecbe6b2258d22

micahswitzer commented on Scones/glob.py:
```diff
@@ -0,0 +1,34 @@
+def glob(env, root: str, extensions: list):
```
This is *very* close to the builtin `Glob` function. Perhaps we could call ours `RecursiveGlob`?

Kfeavel replied:
Addressed in a9bfd261b36f4f272ff1412e79cecbe6b2258d22

Kfeavel committed:
Rename glob tool functions
